### ISSUE TITLE
Support for ecdsaWithSHA512 signature algorithm

### ISSUE THREAD

**Issue by chris-henderson-alation** [rustls/rustls#1367]
Howdy folks,

For better-or-for-worse, our team is forced into working with a certificate authority which is using `ecdsaWithSHA512` and we are getting hit by [UnsupportedSignatureAlgorithm](https://github.com/rustls/webpki/blob/febe4090182b62586ca9b6088bcb687abf39a7bf/src/signed_data.rs#L199), resulting in the generic `UnknnownIssuer` showing up in our application.

Is it at all onerous to add `1.2.840.10045.4.3.4` to the list of [supported algorithms](https://github.com/rustls/rustls/blob/3d121b9d6254a4326a9b92a1c40cb002a84f8188/rustls/src/verify.rs#L25)?

**Comment by djc** [comment]
Sounds like it might be reasonable? Want to submit a PR?

**Comment by ctz** [comment]
Note that that list is a complete list of everything supported by webpki, and thence *ring*.

**Comment by djc** [comment]
Right, so you may have to wait for pluggable crypto provider support (which is coming in the next release). See #521.

**Comment by chris-henderson-alation** [comment]
> Sounds like it might be reasonable? Want to submit a PR?

If only it were as simple as an addition to a list :laughing: I had indeed feared that this was a gap in the backend.

https://github.com/rustls/rustls/issues/521 sounds like a decent avenue forward. Ring itself also seems to have some discussions concerning a wider signature suite (https://github.com/briansmith/ring/issues/824).

**Comment by djc** [comment]
Yes, unfortunately *ring* maintenance seems to be mostly dead for the past two years, so it's unlikely there will be movement on its side soon.

**Comment by cpu** [comment]
I think [aws-lc-rs](https://github.com/aws/aws-lc-rs) has broader [signature algorithm support](https://docs.rs/aws-lc-rs/latest/aws_lc_rs/signature/index.html#statics). Since it's API compatible with *ring* you can use it with webpki today with just a Cargo patch as long as you don't [require `no_std` support](https://github.com/aws/aws-lc-rs/blob/main/aws-lc-rs/README.md#ring-compatibility). That might be an option?

**Comment by cpu** [comment]
> we are getting hit by [UnsupportedSignatureAlgorithm](https://github.com/rustls/webpki/blob/febe4090182b62586ca9b6088bcb687abf39a7bf/src/signed_data.rs#L199), resulting in the generic UnknnownIssuer showing up in our application.

@chris-henderson-alation Separate question: Are you using Rustls v0.21.4+? I expected we would be returning a more specific error in this case in place of `UnknownIssuer` based on some recent improvements.

**Comment by chris-henderson-alation** [comment]
@cpu 

> Since it's API compatible with ring...

That is extremely fascinating. Since this is a transitive dependency through tonic, my go-forward since last week as been to construct a channel using an [openssl](https://github.com/LucioFranco/tonic-openssl) backend.

Although, I was not altogether aware of `aws-lc-rs`, so this is very interesting to catch up on (especially as my org gets more-and-more FIPS scrutiny as of late).

> Are you using Rustls v0.21.4+

`Cargo.lock` indicates 0.21.5.

**Comment by cpu** [comment]
> > Are you using Rustls v0.21.4+
>
> Cargo.lock indicates 0.21.5.

Thanks, I'll dig deeper on that front.

**Comment by cpu** [comment]
> Thanks, I'll dig deeper on that front.

@chris-henderson-alation Could you expand more on how you're using Rustls when you get the `UnknownIssuer` err working with an `ecdsa-with-SHA512` certificate chain?

I tried to reproduce by generating an end entity cert and a CA issuer both using `ecdsa-with-SHA512`. I stood up a Go TLS server using a TLS config serving that EE cert.

When I connect to this server w/ `tlsclient-mio or `simpleclient` I get:
> TLS error: AlertReceived(HandshakeFailure)

and the server side logs:

>  tls: peer doesn't support any of the certificate's signature algorithms

Similarly, looking at `webpki` directly, when I load the `EndEntityCert` and `TlsServerTrustAnchor` issuer and try to use `verify_is_valid_tls_server_cert` on the EE w/ the trust anchor I get the expected `webpki::error::Error::UnsupportedSignatureAlgorithm` error.

The handshake failure could probably be clearer, but it's not an `UnknownIssuer` error like you observed :thinking:

**Comment by cpu** [comment]
Since it seems like https://github.com/rustls/rustls/issues/521 is the right path forward to resolve support for additional signature algorithms I'm going to close this issue out. Following #521 will be the best way to track progress.

Thanks!

**Comment by briansmith** [comment]
P-521 with SHA-512 is in progress in https://github.com/briansmith/ring/pull/1631 and will be in *ring* 0.17.5 or .0.17.6 or so, in a week or two.
